File: leetcode/go/problem152.go
	func maxProduct(nums []int) int {
		mx := nums[0]
		dp := make([][2]int, len(nums))
		dp[0][0] = nums[0]
		dp[0][1] = nums[0]
		for i := 1; i < len(dp); i++ {
			dp[i][0] = max(nums[i], max(dp[i-1][0]*nums[i], dp[i-1][1]*nums[i]))
			dp[i][1] = min(nums[i], min(dp[i-1][0]*nums[i], dp[i-1][1]*nums[i]))
			mx = max(mx, dp[i][0])
		}
		return mx
	}

File: leetcode/python/problem152.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        res = max(nums)

        mn, mx = 1, 1

        for n in nums:
            tmp = mx * n
            mx = max(tmp, n * mn, n)
            mn = min(tmp, n * mn, n)
            res = max(res, mx)
        return res
